[V1] Refactor Structured Output for multiple backends

russellb opened the pull request:
This change does some refactoring of the V1 structured output
implementation to prepare for supporting multiple backends. This code is
already successfully in use in a branch to support a second backend. I
think it will be easier to review other backends if the refactoring goes
in first on its own.

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
[V1] Refactor Structured Output for multiple backends

This change does some refactoring of the V1 structured output
implementation to prepare for supporting multiple backends. This code is
already successfully in use in a branch to support a second backend. I
think it will be easier to review other backends if the refactoring goes
in first on its own.

Signed-off-by: Russell Bryant <rbryant@redhat.com>

WoosukKwon commented:
@aarnphm Could you please take a look?

aarnphm commented on vllm/v1/structured_output/__init__.py:
```diff
@@ -7,104 +7,58 @@
 
     def __init__(self, vllm_config: VllmConfig):
-        self.vocab_size = vllm_config.model_config.get_vocab_size()
+        self.backend: Optional[StructuredOutputBackend] = None
```
fwiw I don't think backend should be None.

Backend should be server-side, idk if changing backend per request is beneficial atm.

russellb replied:
Right - changing per request is explicitly blocked.

It's None here to defer initialization until it's needed for the first time.

russellb replied:
> Right - changing per request is explicitly blocked.

actually I think this part is in my commit adding the next backend, but either way, agree we shouldn't support it

aarnphm commented on vllm/v1/structured_output/__init__.py:
```diff
@@ -114,28 +68,8 @@ def _async_create_grammar(self, request: Request) -> Grammar:
-            vocab_size=self.vocab_size,
-            ctx=ctx,
-        )
+        assert self.backend is not None
```
we should try not to use assert in critical path (and I believe this is)

Given that `-O` and `-OO` will strip assert (ik that we aren't using it atm, but probably worth knowing)

russellb replied:
well, it's something that should never happen and we'd want to know if it did because we know it'll break anyway. It also gives hints to mypy, which is often how I end up adding it.

This should be covered in a style guide somewhere so we have guidelines for the project.

aarnphm commented on vllm/v1/structured_output/__init__.py:
```diff
@@ -147,14 +81,19 @@ def grammar_bitmask(
             return None
 
+        if self._grammar_bitmask is None:
+            assert self.backend is not None
```
ditto

aarnphm commented on vllm/v1/structured_output/backend_types.py:
```diff
@@ -0,0 +1,48 @@
+from abc import ABC, abstractmethod
+
+
+class StructuredOutputOptions(enum.Enum):
```
a bit nit, but can we move this to `backends/`, then it can be `backends/abstract.py`, given that `types` is an actual library in python :)

russellb replied:
but then we'd have `backends/xgrammar.py`, clashing with `xgrammar` itself.  shrug

aarnphm replied:
😮‍💨 makes sense

aarnphm commented on vllm/v1/structured_output/backend_xgrammar.py:
```diff
@@ -0,0 +1,141 @@
+# SPDX-License-Identifier: Apache-2.0
+
+from dataclasses import dataclass, field
```
same here, to `backends/xgrammar.py`

russellb replied:
commented above on this

aarnphm commented on vllm/v1/structured_output/backend_xgrammar.py:
```diff
@@ -0,0 +1,141 @@
+            model_config=vllm_config.model_config,
+            scheduler_config=vllm_config.scheduler_config,
+            parallel_config=vllm_config.parallel_config,
+            lora_config=vllm_config.lora_config)  # type: ignore[arg-type]
```
```suggestion
            lora_config=vllm_config.lora_config,    # type: ignore[arg-type]
        )
```

aarnphm replied:
oh yeah prep for #14485, but not sure we will ever get this in soon.

aarnphm replied:
yeah we can just ignore this for now.

aarnphm commented on vllm/v1/structured_output/backend_xgrammar.py:
```diff
@@ -0,0 +1,141 @@
+            lora_config=vllm_config.lora_config)  # type: ignore[arg-type]
+        tokenizer_group.ping()
+        self.vllm_config = vllm_config
+        self.vocab_size = vllm_config.model_config.get_vocab_size()
```
hmm, https://github.com/vllm-project/vllm/pull/14630

aarnphm commented on vllm/v1/structured_output/backend_xgrammar.py:
```diff
@@ -0,0 +1,141 @@
+        elif request_type == StructuredOutputOptions.REGEX:
+            ctx = self.compiler.compile_regex(grammar_spec)
+        else:
+            logger.error(
```
well this seems a bit redudant, given that we already raise the exception bellow?

russellb replied:
right now I'm not sure this shows up in the log otherwise, i'd have to check

aarnphm reviewed (approved):
just a few comments wrt structuring, otherwise LGTM

WoosukKwon commented:
@aarnphm @russellb Just wanted to double check: Is this PR ready for merge?

aarnphm commented:
yes, but I don't think this would block 0.8.0

aarnphm commented on vllm/v1/structured_output/__init__.py:
```diff
@@ -7,104 +7,58 @@
+        #
+        # NOTE: We only support a single backend. We do NOT support different
+        # backends on a per-request basis in V1 (for now, anyway...).
+        if self.backend is None:
+            backend_name = request.sampling_params.guided_decoding.backend_name
+            if backend_name == "xgrammar":
+                self.backend = XgrammarBackend(self.vllm_config)
+            else:
+                raise ValueError(
+                    f"Unsupported structured output backend: {backend_name}")
```
@russellb we should use `decoding_config` instead of request sampling_params.

russellb replied:
good catch - I actually fixed this in #14779, but it should be in this one

russellb replied:
fixed

russellb committed:
Handle backend choice more gracefully

- Ensure request-level choice matches server side config if specified.
  We don't support requesting something different than what was
  configured.

- Fix handling when a backend is not specified in the request.

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb commented:
This is good now -- I also hit a different bug along the way and fixed it in #14826

WoosukKwon commented:
@russellb Just to double check: This is not a release blocker, is it?

aarnphm commented:
This is not.

aarnphm commented on vllm/v1/structured_output/backend_types.py:
```diff
@@ -0,0 +1,48 @@
+    """Request-level backend for structured output requests."""
+
+    @abstractmethod
+    def accept_tokens(self, request_id: str, tokens: list[int]) -> bool:
```
nit: we can add docstring here instead of `pass` so that documentHover works for all subclass.

russellb replied:
done

russellb committed:
Add docstrings for structured output abstract classes

Signed-off-by: Russell Bryant <rbryant@redhat.com>

DarkLight1337 commented on vllm/v1/structured_output/__init__.py:
```diff
@@ -7,104 +7,58 @@
     import numpy as np
     import numpy.typing as npt
-    import xgrammar as xgr
+    import torch
```
Why do we need to lazy import these in the first place? I think only `Request` needs to be lazy imported

aarnphm replied:
Because we only use this for typing, there is no need to reimport this.

fwiw we might want to lazy import torch until engine serve at some point.

Right now `import vllm` is relatively slow.

DarkLight1337 reviewed (approved):
Otherwise LGTM

NickLucche reviewed (approved):
For sanity check I've also ran this on V1 TPU and it's not interfering on start up

russellb commented:
> For sanity check I've also ran this on V1 TPU and it's not interfering on start up

thank you for checking! much appreciated
